internal/informers/core_filteredsecrets.go

```diff
@@ -0,0 +1,333 @@
+	typedLister           corev1listers.SecretLister
+	typedClient           typedcorev1.SecretsGetter
+	// Go recommends to not store context in
+	// structs, but here we have no other way as we need to use root context inside
+	// Get whose signature is defined upstream and does not accept context
+	ctx context.Context
+}
+
+func (snl *secretNamespaceLister) Get(name string) (*corev1.Secret, error) {
+	log := logf.FromContext(snl.ctx)
+	log = log.WithValues("secret", name, "namespace", snl.namespace)
+
+	var secretFoundInTypedCache, secretFoundInMetadataCache bool
+	secret, typedCacheErr := snl.typedLister.Secrets(snl.namespace).Get(name)
+	if typedCacheErr == nil {
+		secretFoundInTypedCache = true
+	}
+
+	if typedCacheErr != nil && !apierrors.IsNotFound(typedCacheErr) {
+		log.Error(typedCacheErr, "error getting secret from typed cache")
+		return nil, fmt.Errorf("error retrieving secret from the typed cache: %w", typedCacheErr)
+	}
+	_, partialMetadataGetErr := snl.partialMetadataLister.Namespace(snl.namespace).Get(name)
+	if partialMetadataGetErr == nil {
+		secretFoundInMetadataCache = true
+	}
+
+	if partialMetadataGetErr != nil && !apierrors.IsNotFound(partialMetadataGetErr) {
+		return nil, fmt.Errorf("error retrieving object from partial object metadata cache: %w", partialMetadataGetErr)
+	}
+
+	if secretFoundInMetadataCache {
+		// if secret is found in both caches log an error and return the version from kube apiserver
+		if secretFoundInTypedCache {
+			key := types.NamespacedName{Namespace: snl.namespace, Name: name}
+			log.Info(fmt.Sprintf("warning: possible internal error: stale cache: secret found both in typed cache and in partial cache: %s", pleaseOpenIssue), "secret", key)
+		}
+		return snl.typedClient.Secrets(snl.namespace).Get(snl.ctx, name, metav1.GetOptions{})
+	}
+
+	if secretFoundInTypedCache {
+		return secret, nil
+	}
+
+	// If we get here it is because secret was found neither in typed cache
+	// nor partial metadata cache
+	return nil, apierrors.NewNotFound(schema.GroupResource{Group: corev1.GroupName, Resource: corev1.ResourceSecrets.String()}, name)
+}
+
+func (snl *secretNamespaceLister) List(selector labels.Selector) ([]*corev1.Secret, error) {
+	log := logf.FromContext(snl.ctx)
+	log = log.WithValues("secrets namespace", snl.namespace, "secrets selector", selector.String())
+	matchingSecretsMap := make(map[types.NamespacedName]*corev1.Secret)
+	typedSecrets, err := snl.typedLister.List(selector)
+	if err != nil {
+		log.Error(err, "error listing Secrets from typed cache")
+		return nil, err
+	}
+	for _, secret := range typedSecrets {
+		key := types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}
+		matchingSecretsMap[key] = secret
+	}
+	metadataSecrets, err := snl.partialMetadataLister.List(selector)
+	if err != nil {
+		log.Error(err, "error listing Secrets from metadata only cache")
+		return nil, err
+	}
+
+	// In practice this section will likely never be used. The only place
```

Comment:
can we also add a warning log here (eg. in case len(metadataSecrets) > 10) that warns the user that there might be performance degradation due to unexpected usage patterns and that they should open a GH issue?